### ISSUE TITLE
Git Workflow

### ISSUE THREAD

**Issue by eileenxue** [eileenxue/lhl-midterm#1]
Here's the git workflow for our team to avoid any merge conflicts or complex issues.

### Daily Starting Point
- We should always start off the latest version of the master branch every day. 
- We will review the master branch together and fix any small issues that may have appeared during recent merges.
- Be on your master branch on your own computer and then do a `git pull` to pull in the latest stuff from the remote master
- There are two ways you can start your daily work: 
  - You can create a NEW branch to work on after pulling in the master 
    - `git checkout -b feature/mybranch` (the -b means create and checkout means switch to that branch)
  - OR if you really want to continue working on your existing branch, you can update that branch to have the latest changes 
    - `git checkout feature/existingbranch` and then `git merge master` 
    - Another method `git checkout feature/existingbranch` and then `git rebase master` ONLY after you have already merged your own branch into the current master

### Committing Your Work
- You should add the files and commit your work on your branches as often as you need to. It will be stored locally. 
  - `git add . ` (You have to be in your main folder for it to work)
  - `git commit -m "Your commit message here"`
- **IMPORTANT**: Do not use the command `git push --all` while on your own branches because you'll also end up pushing your master branch that may not be up to date yet.

### Merging Your Work
- Once your work on your branches is ready to be merged into master, follow the following steps:
  - Save all your edited files, then add and commit them first on your own branch.
  - Switch to the master branch `git checkout master`. Don't worry if your files update or revert back to before you started working, it's still on your local branch! 
  - Do a `git pull` to make your local master branch be the same as the remote master branch, now you have the latest master across the entire team. If other people made new changes/merges, you'll have it locally on your computer.
  - Now it's time to merge YOUR new changes into the master. You can double-check you're on the right branch by using `git status` or `git branch` to see all your branches and see which one is highlighted.
  - While on the master branch, do `git merge your-branch-name` and if all is successful, it should work without issues. You may encounter the vim editor asking why you want to merge. Use this [answer](https://stackoverflow.com/a/19085954/5566317) to guide you through.
  - Now do a `git push` to push your new local master with your branch changes to the remote master copy so that everyone can see it and subsequently pull from it. You may also use `git push --all` at this point to also push your local branches up to the repo. 

### Conflicts with Merges
- To be updated
